src/timepicker/timepicker.ts

```diff
@@ -224,7 +224,11 @@ export class NgbTimepicker implements ControlValueAccessor,
   updateHour(newVal: string) {
     const isPM = this.model.hour >= 12;
     const enteredHour = toInteger(newVal);
-    this.model.updateHour((this.meridian ? enteredHour % 12 : enteredHour) + (this.meridian && isPM ? 12 : 0));
+    if (this.meridian && (isPM && enteredHour < 12 || !isPM && enteredHour === 12)) {
+      this.model.updateHour(enteredHour + 12);
```

Comment:
if you enter `30` it will try to set it as an hour without any corrections and `updateHour` will do `% 24`, so the result will be `6`, so it will switch to AM.